### PR TITLE
Fix : wrong virtual stock when partial reception

### PR DESCRIPTION
**Pull request by atm-maxime**
Virtual stock was not decreasing qty already received on open supplier order

### CODE REVIEW

File: htdocs/product/class/product.class.php [hunk 1]
```diff
@@ -3477,7 +3477,7 @@ function load_virtual_stock()
             if ($result < 0) dol_print_error($this->db,$this->error);
             $stock_commande_fournisseur=$this->stats_commande_fournisseur['qty'];
 
-            $result=$this->load_stats_reception(0,'4');
+            $result=$this->load_stats_reception(0,'3,4');
```
Comment (eldy) [inline htdocs/product/class/product.class.php:3480]: Status 3 is "process running" nothing received yet.
4 is "Received partially".
May be the problem is that with 3.9 (and may be other version), when you record one reception, the status of the supplier order is not set from 3 to 4 ? Can you check that ?

**Comment by BebZ** [comment]
Thumbs up to point that :)

These statuses are set after a stock dispatch... manually with a text field linked
"qty already received on open supplier order" will be decreased from virtual stock ?

Is the status 5 "received completely" ?

Is it possible to add a status : 6 received completly but with conflicts ? 
for example on quantities delivered (damaged, broken lost broken or else....)

**Comment by eldy** [comment]
Yes, status 5 is "Received completely"
See  https://wiki.dolibarr.org/index.php/Module_Suppliers_Orders

Currently, this is a manual step, but for moving from 3 to 4, there is no reason to have this done manually.

**Comment by aljawaid** [comment]
> Is it possible to add a status : 6 received completly but with conflicts ?
for example on quantities delivered (damaged, broken lost broken or else....)

That would be very helpful as some stock can be delivered to us damaged but we still accept the order.

**Comment by BebZ** [comment]
Theres is already an inexploited status 6 ... no ? 

image

https://wiki.dolibarr.org/index.php/Module_Suppliers_Orders#Cl.C3.B4turer_une_commande_fournisseur

**Comment by atm-maxime** [comment]
I thought that status 3 was "ordered - waiting for reception" and statuses 4 and 5 (partially and completely received) were final. I set my order to 4 or 5 when i am not waiting for another reception. You say that after the first reception, status should automatically go to 4, and must be set to 5 if everything is received ?

We can add that, but for now, reception can be made on a status 3 order, so this status should be used to check for reception in the virtual stock calculation.

**Comment by BebZ** [comment]
For virtual I agree with atm not ventilated or not yet received products shouldn't be taken in the sum of virtual products (future products we will receive) as it is done for customers shipments (future products that we will send)
I set to 4 when it is a partial reception but it can be final / or not 
We must know if there are missing/broken products even if supplier will not send them again.
I set to 5 even if all products were not received when we have solved the problem of missing / broken products. Maybe not the right way ?

Maybe the right way would be to add a status or use 6th status to say received completely but with problem of missing /too much/broken products.

**Comment by BebZ** [comment]
So, what about adding a new status for supplier orders ? 
I see number 8 is still "free" ? can we use it ?

**Comment by BebZ** [comment]
up ? !

**Comment by BebZ** [comment]
up² ?

**Pull request closed without merging**